google/cloud/bigquery/table.py

```diff
@@ -1891,10 +1891,21 @@ def interval(self, value):
     def _key(self):
         return tuple(sorted(self._properties.items()))
 
+    def __eq__(self, other):
+        if not isinstance(other, PartitionRange):
+            return NotImplemented
+        return self._key() == other._key()
+
+    def __ne__(self, other):
+        return not self == other
+
     def __repr__(self):
         key_vals = ["{}={}".format(key, val) for key, val in self._key()]
         return "PartitionRange({})".format(", ".join(key_vals))
 
+    def __hash__(self):
+        return hash(self._key())
```

Comment:
A hash should never change during the object's lifetime, as it can lead to unexpected results. Mutable objects should thus not be hashable:
```py
>>> from google.cloud.bigquery import PartitionRange
>>> pr = PartitionRange(start=42)
>>> s = {pr}
>>> pr in s
True
>>> pr.start = 43
>>> s
{PartitionRange(start=43)}
>>> pr in s
False   # huh??
```
Until Python 2 compatibility needs to be preserved, we need to manually mark the object as unhashable:
```py
__hash__ = None
```

Reply:
I have tried same with the `table.TimePartitioning` and `dataset.EncryptionConfiguration` classes and got the output same as you had mentioned above, any suggestion on that?

Reply:
The "same output" meaning the same issue when using the instances in a hashable collection?

If an object with a custom `eq` is mutable, it really shouldn't be hashable, unless the custom equality and hash are based on some immutable subset of the attributes. If we want hashability, we need to make the instances immutable. If we nevertheless still want to "edit" such instances, we can add methods that make a _copy_ of the instance containing the changes (IIRC the Retry objects from API core are an example of that).

It might not be worth it, though, depending on how much we really need hashability.